contracts/protocol/facets/ExchangeHandlerFacet.sol

```diff
@@ -804,6 +804,40 @@ contract ExchangeHandlerFacet is IBosonExchangeHandler, BuyerBase, DisputeBase {
                     twinReceipt.amount = twin.amount;
                     twinReceipt.tokenType = twin.tokenType;
 
+                    if (twin.tokenType == TokenType.NonFungibleToken) {
+                        // Get all ranges of twins that belong to the seller and to the same token address of the new twin to validate if range is available
+                        TokenRange[] storage twinRanges = lookups.twinRangesBySeller[seller.id][twin.tokenAddress];
+
+                        uint256 lastIndex = twinRanges.length - 1;
+
+                        // Find the range that contains the twin
```

Comment:
Looping does the job, but did you also consider introducing a new mapping `rangeIdByTwinId`?
If we store that during the twin creation, we could more easily find the associated range here.

And `removeTwin` could be optimized using the same mapping as well.

If you'll think about it, then also think how can we back-fill the data for existing twins (if they exists). Probably `initV2_3_0` could accept an array of key-value pairs and it could fill `rangeIdByTwinId` for existing twins.

Reply:
I added the mapping
I'll work tomorrow on how to back-fill data